File: LeetCode - Top Interview 150/222.py
# 222. Count Complete Tree Nodes
# https://leetcode.com/problems/count-complete-tree-nodes/


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        def get_leftmost_height(node):
            result = 0
            current = node

            while current:
                result += 1
                current = current.left

            return result

        def get_rightmost_height(node):
            result = 0
            current = node

            while current:
                result += 1
                current = current.right

            return result

        def get_count(node):
            result = 0

            queue = []
            queue.append(node)

            while queue:
                current = queue.pop(0)
                result += 1

                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)

            return result

        leftmost_height = get_leftmost_height(root)
        rightmost_height = get_rightmost_height(root)

        if leftmost_height == rightmost_height:
            return (2**leftmost_height) - 1
        else:
            return get_count(root)

File: LeetCode - Algorithm/Binary Tree/222.js
// 222. Count Complete Tree Nodes
// https://leetcode.com/problems/count-complete-tree-nodes/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function (root) {
  let result = 0;

  const queue = [];
  if (root) queue.push(root);

  while (queue.length > 0) {
    const node = queue.shift();
    result += 1;

    if (node.left) queue.push(node.left);
    if (node.right) queue.push(node.right);
  }

  return result;
};
